### DISCUSSION TITLE
Toggle missing and known traits for properties and spaces

### DISCUSSION THREAD

**Discussion by marswill** [pi-base/data#555] [Ideas]
Right now, when you look at the properties tab for a space, you don't see the unknown properties, and have to infer what they are by crosschecking with the main list in another tab or window.

I took a stab at changing this in a [recent pull request](https://github.com/pi-base/web/pull/133), which allows users to toggle missing and known traits on or off on the properties tab of space (and similarly, for the spaces tab of a property, though this is less exciting since you can already get that info from the explore page).

Would be grateful for any feedback.  Here is what it looks like:

https://missing-traits.topology.pages.dev/spaces/S000001

**Comment by prabau** [comment]
I like this.  Looks like it could be useful.

Some comments about the interface:
- I think the default when accessing a page (space or property) should be to have to have the unknown properties hidden.  Ordinary users won't care about seeing it.
- In the old interface, the asserted properties were always shown and one could toggle the deduced properties.  In the new interface, there are now separate buttons for toggling the asserted and the deduced.  But it seems to behave a little funny: when I click on the asserted button, the deduced properties also disappear, but the deduced button is still labeled as "Hide" even though there is nothing to hide.  So the toggling buttons for asserted and deduced seem they are independent, but they are not.  I am not sure what behavior we want, but there is something to play with and decide.
- The buttons are listed in the order (asserted, unknown, deduced).  I think it would be better to follow the order (asserted, deduced, unknown), as unknown is least important.

**Comment by StevenClontz** [comment]
Hey this is something I've wanted to tackle for a while, thanks for taking a stab! We need better discoverability of "missing" data.

No time right now for careful feedback but thanks for moving this forward.

**Comment by marswill** [comment]
> * I think the default when accessing a page (space or property) should be to have to have the unknown properties hidden.  Ordinary users won't care about seeing it.

Seems reasonable.

> * In the old interface, the asserted properties were always shown and one could toggle the deduced properties.  In the new interface, there are now separate buttons for toggling the asserted and the deduced.  But it seems to behave a little funny: when I click on the asserted button, the deduced properties also disappear, but the deduced button is still labeled as "Hide" even though there is nothing to hide.  So the toggling buttons for asserted and deduced seem they are independent, but they are not.  I am not sure what behavior we want, but there is something to play with and decide.

Yeah, technically right now the way I programmed this was that the buttons are for showing and hiding "unknown", "known", and "deduced", which made sense to me in my head at the time, but after reading your comment, I'm realizing that's more than a little weird, and a delineation between "asserted, deduced, unknown" makes much more sense. After all, these are the three disjoint categories that people care about.

> * The buttons are listed in the order (asserted, unknown, deduced).  I think it would be better to follow the order (asserted, deduced, unknown), as unknown is least important.

Agreed.

**Comment by prabau** [comment]
> Yeah, technically right now the way I programmed this was that the buttons are for showing and hiding "unknown", "known", and "deduced", which made sense to me in my head at the time, but after reading your comment, I'm realizing that's more than a little weird, and a delineation between "asserted, deduced, unknown" makes much more sense. After all, these are the three disjoint categories that people care about.

I guess I did not realize that the first button was for "known", I assumed it was for "asserted".  So we could use the "human" icon for this "asserted" button.

**Comment by marswill** [comment]
I've added a replacement for the 404 page, instead inviting contributors to supply missing properties.  I had a random thought that maybe it would be nice to show the related theorems for a property when inviting contributions.  After looking at the result I think maybe that was overkill, but I left it in for now to see what others think. 

https://missing-traits.topology.pages.dev/spaces/S000092/properties/P000005

**Comment by prabau** [comment]
> show the related theorems for a property when inviting contributions

It looks nice, but maybe overkill.  Giving the list of all theorems related to the property in question seems to dilute the message.

We should discuss in the next zoom meeting.

**Comment by StevenClontz** [comment]
Thanks for showing this off today! In around a week or so I hope to have some time to review the PR properly and maybe help with test coverage if necessary.

**Comment by marswill** [comment]
Based on @StevenClontz's suggestion (which I didn't fully process the usefulness of during the meeting but I realize now it's obviously helpful), I've added filtering based on true/false values of traits.

In order to not have a million competing boolean flags that overlap with each other (e.g., you don't wan't to have to turn off the true ones and turn on the question marks to see the missing ones), I've just got a single button for each conceivable display mode that could be of interest:

- Show everything
- Show all known traits
- Show all user asserted traits
- Show all true traits
- Show all false traits
- Show all missing traits

This way what you want is just one click away.

~I've also made "Show all true traits" the default mode.  This seems the most intuitive to me - when I was first using the site I remember being thrown for a loop sometimes clicking on a space and looking down at the list wondering what all those blatantly incorrect properties were doing there, then having to eventually realize there's a little X off to the side.~

Based on feedback in the PR and reconsidering, default mode is now "Show all known traits".
